src/test/socket/ioctl/test_ioctl.rs

```diff
@@ -184,3 +194,112 @@ fn test_fionread(init_method: SocketInitMethod, sock_type: libc::c_int) -> Resul
+        // neither socket has received any data, so should return ENOENT for inet sockets
+        let expected_result = match (init_method.domain(), sock_type) {
+            (libc::AF_INET, _) => Err(libc::ENOENT),
+            (libc::AF_UNIX, _) => Err(libc::ENOTTY),
+            _ => unimplemented!(),
+        };
+        test_utils::result_assert_eq(
+            ioctl_siocgstamp(fd_client),
+            expected_result,
+            "Unexpected SIOCGSTAMP result",
+        )?;
+        test_utils::result_assert_eq(
+            ioctl_siocgstamp(fd_peer),
+            expected_result,
+            "Unexpected SIOCGSTAMP result",
+        )?;
+
+        // send data from the client to the peer
+        let flags = nix::sys::socket::MsgFlags::empty();
+        nix::sys::socket::send(fd_client, &[1, 2, 3], flags).unwrap();
+
+        // approximately the time that we sent the message
+        let send_time = std::time::SystemTime::now()
+            .duration_since(std::time::SystemTime::UNIX_EPOCH)
+            .unwrap();
+
+        // shadow needs to run events, but we also sleep so that we make the recv() call much later
+        // than the send() call
+        std::thread::sleep(Duration::from_millis(50));
+
+        // use use a small threshold when comparing the send time to the recv time below since the
+        // message is only travelling over localhost; should be much shorter than the sleep above
+        let threshold = Duration::from_millis(1);
+
+        // recv() has not been called on the socket, so should still return ENOENT for inet sockets
+        let expected_result = match (init_method.domain(), sock_type) {
+            (libc::AF_INET, _) => Err(libc::ENOENT),
+            (libc::AF_UNIX, _) => Err(libc::ENOTTY),
+            _ => unimplemented!(),
+        };
+        test_utils::result_assert_eq(
+            ioctl_siocgstamp(fd_peer),
+            expected_result,
+            "Unexpected SIOCGSTAMP result",
+        )?;
+
+        // receive data at the peer
+        let flags = nix::sys::socket::MsgFlags::empty();
+        nix::sys::socket::recv(fd_peer, &mut [0u8; 3], flags).unwrap();
+
+        // check the result of SIOCGSTAMP on the peer; only supported by udp sockets
+        let expected_err = match (init_method.domain(), sock_type) {
+            (libc::AF_INET, libc::SOCK_DGRAM) => None,
+            (libc::AF_INET, _) => Some(libc::ENOENT),
+            (libc::AF_UNIX, _) => Some(libc::ENOTTY),
+            _ => unimplemented!(),
+        };
+        match expected_err {
+            None => {
+                // the receive time reported by the kernel
+                let recv_time = ioctl_siocgstamp(fd_peer).unwrap();
+                let recv_time = Duration::from_secs(recv_time.tv_sec.try_into().unwrap())
+                    + Duration::from_micros(recv_time.tv_usec.try_into().unwrap());
+
+                // get the difference (we can't know if the receive or send time will be smaller)
```

Comment:
How could the receive time be before the send time?

Reply:
I expanded the comment here. We measure the send time as soon as the `send()` call returns, but the message may have already been received before we can take that measurement. For example:

1. `send()` call
  a. since it's localhost, the packet is given directly to the receiving socket within the `send()` call
  b. receive time is recorded by the kernel here
2. shadow records the send time with `let send_time = std::time::SystemTime::now()`

In this case the send time is later than the receive time.